realm/realm-library/src/main/java/io/realm/BaseRealm.java

```diff
@@ -637,47 +533,61 @@ protected static synchronized void validateAgainstExistingConfigurations(RealmCo
+            @Override
+            public void onResult(int count) {
+                if (count != 0) {
+                    throw new IllegalStateException("It's not allowed to delete the file associated with an open Realm. " +
+                            "Remember to close() all the instances of the Realm before deleting its file.");
+                }
 
-        boolean realmDeleted = true;
-        String canonicalPath = configuration.getPath();
-        File realmFolder = configuration.getRealmFolder();
-        String realmFileName = configuration.getRealmFileName();
-        List<File> filesToDelete = Arrays.asList(new File(canonicalPath),
-                new File(realmFolder, realmFileName + ".lock"),
-                new File(realmFolder, realmFileName + ".lock_a"),
-                new File(realmFolder, realmFileName + ".lock_b"),
-                new File(realmFolder, realmFileName + ".log"));
-        for (File fileToDelete : filesToDelete) {
-            if (fileToDelete.exists()) {
-                boolean deleteResult = fileToDelete.delete();
-                if (!deleteResult) {
-                    realmDeleted = false;
-                    RealmLog.w("Could not delete the file " + fileToDelete);
+                String canonicalPath = configuration.getPath();
+                File realmFolder = configuration.getRealmFolder();
+                String realmFileName = configuration.getRealmFileName();
+                List<File> filesToDelete = Arrays.asList(new File(canonicalPath),
+                        new File(realmFolder, realmFileName + ".lock"),
+                        new File(realmFolder, realmFileName + ".lock_a"),
+                        new File(realmFolder, realmFileName + ".lock_b"),
+                        new File(realmFolder, realmFileName + ".log"));
+                for (File fileToDelete : filesToDelete) {
+                    if (fileToDelete.exists()) {
+                        boolean deleteResult = fileToDelete.delete();
+                        if (!deleteResult) {
+                            realmDeleted[0] = false;
+                            RealmLog.w("Could not delete the file " + fileToDelete);
+                        }
+                    }
                 }
             }
-        }
+        });
 
-        return realmDeleted;
+        return realmDeleted[0];
     }
 
     /**
      * Compacts the Realm file defined by the given configuration.
      */
-    public static synchronized boolean compactRealm(RealmConfiguration configuration) {
+    public static boolean compactRealm(final RealmConfiguration configuration) {
         if (configuration.getEncryptionKey() != null) {
             throw new IllegalArgumentException("Cannot currently compact an encrypted Realm.");
         }
 
-        if (isFileOpen(configuration)) {
-            throw new IllegalStateException("Cannot compact an open Realm");
-        }
+        final boolean[] result = {false};
```

Comment:
I would replace this with an AtomicBoolean instead. It expresses the intent better IMO.